File: python_backend/src/api/v1/content.py
"""
Content Agent API Routes
"""
import logging
import json
from typing import AsyncGenerator

from fastapi import APIRouter
from fastapi.responses import StreamingResponse

from ...agents.content_strategist_agent import (
    content_strategist_chat,
    ChatStrategistRequest,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/content", tags=["content"])


@router.post("/strategist/chat")
async def chat_strategist(request_body: ChatStrategistRequest):
    """
    POST /api/v1/content/strategist/chat
    
    Stream chat with content strategist agent.
    Memory handled automatically via thread_id.
    """
    
    async def generate_stream() -> AsyncGenerator[str, None]:
        full_response = ""
        
        async for chunk in content_strategist_chat(request_body):
            step = chunk.get("step", "")
            content = chunk.get("content", "")
            
            if step == "error":
                yield f"data: {json.dumps({'type': 'error', 'message': content})}\n\n"
                return
            
            if content:
                full_response = content
                yield f"data: {json.dumps({'type': 'update', 'step': step, 'content': content})}\n\n"
        
        yield f"data: {json.dumps({'type': 'done', 'response': full_response})}\n\n"
    
    return StreamingResponse(
        generate_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        }
    )

File: src/lib/python-backend/api/content.ts
/**
 * Content Agent API
 * 
 * Streaming chat with the Python backend content strategist agent.
 * Supports multimodal input (text, images, PDFs).
 */

import { ENDPOINTS, getEndpointUrl } from '../config';
import { ContentBlock } from '@/lib/multimodal-utils';

export interface ChatRequest {
    message: string;
    threadId: string;
    modelId?: string;
    /** Content blocks for multimodal input (images, PDFs) */
    contentBlocks?: ContentBlock[];
}

export interface StreamEvent {
    type: 'update' | 'done' | 'error';
    step?: string;
    content?: string;
    response?: string;
    message?: string;
}

/**
 * Chat with content strategist (streaming)
 * 
 * Opens SSE connection and streams response.
 * Supports multimodal input via contentBlocks.
 * 
 * @param request - Chat request with message, threadId, and optional contentBlocks
 * @param onUpdate - Callback for each content update
 * @param onComplete - Callback when stream completes
 * @param onError - Callback for errors
 */
export async function chatStrategist(
    request: ChatRequest,
    onUpdate: (content: string) => void,
    onComplete: (response: string) => void,
    onError: (error: Error) => void
): Promise<void> {
    const url = getEndpointUrl(ENDPOINTS.content.chat);

    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'text/event-stream',
            },
            body: JSON.stringify(request),
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const reader = response.body?.getReader();
        if (!reader) {
            throw new Error('No response body');
        }

        const decoder = new TextDecoder();
        let buffer = '';
        let fullResponse = '';

        while (true) {
            const { done, value } = await reader.read();

            if (done) break;

            buffer += decoder.decode(value, { stream: true });

            // Process complete SSE messages
            const lines = buffer.split('\n');
            buffer = lines.pop() || '';

            for (const line of lines) {
                if (line.startsWith('data: ')) {
                    try {
                        const jsonStr = line.slice(6);
                        console.log('[Stream] Received:', jsonStr);
                        const data: StreamEvent = JSON.parse(jsonStr);

                        if (data.type === 'update' && data.content) {
                            fullResponse = data.content;
                            console.log('[Stream] Update:', data.content.substring(0, 100) + '...');
                            onUpdate(data.content);
                        } else if (data.type === 'done') {
                            console.log('[Stream] Done:', data.response?.substring(0, 100) || fullResponse.substring(0, 100) + '...');
                            onComplete(data.response || fullResponse);
                            return;
                        } else if (data.type === 'error') {
                            console.error('[Stream] Error:', data.message);
                            throw new Error(data.message || 'Stream error');
                        }
                    } catch (parseError) {
                        console.warn('[Stream] Parse error:', parseError, 'Line:', line);
                    }
                }
            }
        }

        onComplete(fullResponse);

    } catch (error) {
        onError(error instanceof Error ? error : new Error(String(error)));
    }
}

/**
 * Create a unique thread ID
 */
export function createThreadId(): string {
    return `thread-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
}
